hare/flows_test.go

```diff
@@ -396,7 +241,8 @@ func Test_multipleCPs(t *testing.T) {
 	finalLyr := types.GetEffectiveGenesis().Add(totalCp)
 	test := newHareWrapper(totalCp)
 	totalNodes := 10
-	cfg := config.Config{N: totalNodes, F: totalNodes / 2, WakeupDelta: time.Second, RoundDuration: 5 * time.Second, ExpectedLeaders: 5, LimitIterations: 1000, LimitConcurrent: 100, Hdist: 20}
+	networkDelay := time.Second
+	cfg := config.Config{N: totalNodes, WakeupDelta: networkDelay, RoundDuration: networkDelay, ExpectedLeaders: 5, LimitIterations: 1000, LimitConcurrent: 100, Hdist: 20}
 
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
```

Comment:
If `cancel` is never explicitly called, `context.WithCancel` isn't needed.

Maybe here `gomock.WithContext` would be a good fit. The context it returns is canceled when any assert fails.

Further down the test when ranging over channels we can add a `select` that checks if the context is `Done` and if yes returns instead of getting stuck on a channel read that will never happen.

This could fix the issues we have with timeouts on this test, if that's the cause?

Reply:
@fasmat do you mind if i leave this for the person (hopefully @piersy) attempting to fix this test?

i don't want to do a half-ass job if i am doing it, but this is just too much change to do it right.
because if you want to do a select later when waiting for termination, that means you have to plug the controller through all mocks. and that affect other tests too because they use the same harness.

Reply:
sure 🙂 I just thought of it when I saw the `WithContext` function in the gomock documentation. I think this could be the issue: some unexpected call is made on a go routine that is not the main -> test gets stuck but doesn't print why.

Reply:
Refactoring this test is out of scope for this PR. There is also another PR here that is already working on it: https://github.com/spacemeshos/go-spacemesh/pull/4100